File: app/src/main/java/com/movieapp/huxymovies/datasource/ResultDataSource.java
package com.movieapp.huxymovies.datasource;

import android.arch.paging.PageKeyedDataSource;
import android.support.annotation.NonNull;

import com.movieapp.huxymovies.datasource.remote.RetrofitClient;
import com.movieapp.huxymovies.model.MovieApiResponse;
import com.movieapp.huxymovies.model.Result;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.movieapp.huxymovies.utils.Utils.INSTANCE;


public class ResultDataSource extends PageKeyedDataSource<Long, Result> {


    private static final long FIRST_PAGE = 1;

    @Override
    public void loadInitial(@NonNull LoadInitialParams<Long> params, @NonNull LoadInitialCallback<Long, Result> callback) {

        RetrofitClient.getInstance()
                .getApi()
                .getMovies(INSTANCE.getAPI_KEY(), FIRST_PAGE, INSTANCE.getLANGUAGE())
                .enqueue(new Callback<MovieApiResponse>() {
                    @Override
                    public void onResponse(Call<MovieApiResponse> call, Response<MovieApiResponse> response) {

                        if (response.body() != null) {

                            callback.onResult(response.body().getMResults(), null, FIRST_PAGE + 1L);


                        }
                    }

                    @Override
                    public void onFailure(Call<MovieApiResponse> call, Throwable t) {

                    }
                });

    }

    @Override
    public void loadBefore(@NonNull LoadParams<Long> params, @NonNull LoadCallback<Long, Result> callback) {


        RetrofitClient.getInstance()
                .getApi()
                .getMovies(INSTANCE.getAPI_KEY(), params.key, INSTANCE.getLANGUAGE())
                .enqueue(new Callback<MovieApiResponse>() {
                    @Override
                    public void onResponse(Call<MovieApiResponse> call, Response<MovieApiResponse> response) {

                        Long key = (params.key > 1) ? params.key - 1 : null;

                        if (response.body() != null) {

                            callback.onResult(response.body().getMResults(), key);

                        }
                    }

                    @Override
                    public void onFailure(Call<MovieApiResponse> call, Throwable t) {

                    }
                });

    }

    @Override
    public void loadAfter(@NonNull LoadParams<Long> params, @NonNull LoadCallback<Long, Result> callback) {


        RetrofitClient.getInstance()
                .getApi()
                .getMovies(INSTANCE.getAPI_KEY(), params.key, INSTANCE.getLANGUAGE())
                .enqueue(new Callback<MovieApiResponse>() {
                    @Override
                    public void onResponse(Call<MovieApiResponse> call, Response<MovieApiResponse> response) {

                        Long key = response.body().getMTotalPages() > response.body().getMPage() ? params.key + 1 : null;

                        if (response.body() != null) {

                            callback.onResult(response.body().getMResults(), key);
                        }
                    }

                    @Override
                    public void onFailure(Call<MovieApiResponse> call, Throwable t) {

                    }
                });
    }
}

File: app/src/main/java/com/movieapp/huxymovies/datasource/ResultDataSource.kt
package com.movieapp.huxymovies.datasource

import android.arch.paging.PageKeyedDataSource
import com.movieapp.huxymovies.HuxyMovies
import com.movieapp.huxymovies.datasource.remote.RetrofitClient
import com.movieapp.huxymovies.model.MovieApiResponse
import com.movieapp.huxymovies.model.Result
import com.movieapp.huxymovies.utils.Utils.API_KEY
import com.movieapp.huxymovies.utils.Utils.LANGUAGE
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


class ResultDataSource : PageKeyedDataSource<Long, Result>() {

    override fun loadInitial(params: PageKeyedDataSource.LoadInitialParams<Long>, callback: PageKeyedDataSource.LoadInitialCallback<Long, Result>) {

        RetrofitClient.instance
                .api
                .getMovies(API_KEY, FIRST_PAGE, LANGUAGE)
                .enqueue(object : Callback<MovieApiResponse> {
                    override fun onResponse(call: Call<MovieApiResponse>, response: Response<MovieApiResponse>) {

                        if (response.isSuccessful) {

                            Thread(Runnable {

                                // First clear the local database.
                                HuxyMovies.database!!.resultsDao().deleteAllMovies()

                                // Then , Insert into the local database.
                                HuxyMovies.database!!.resultsDao().insertAllMovies(response.body()!!.mResults!!)

                                callback.onResult(response.body()!!.mResults!!,
                                        null, FIRST_PAGE + 1L)

                            }).start()

                        }
                    }

                    override fun onFailure(call: Call<MovieApiResponse>, t: Throwable) {
                        Thread(Runnable {
                            callback.onResult(HuxyMovies.database!!.resultsDao().getAllMovies()!!,
                                    null, FIRST_PAGE + 1L)
                        }).start()
                    }
                })

    }

    override fun loadBefore(params: PageKeyedDataSource.LoadParams<Long>, callback: PageKeyedDataSource.LoadCallback<Long, Result>) {


        RetrofitClient.instance
                .api
                .getMovies(API_KEY, params.key, LANGUAGE)
                .enqueue(object : Callback<MovieApiResponse> {
                    override fun onResponse(call: Call<MovieApiResponse>, response: Response<MovieApiResponse>) {

                        val key = if (params.key > 1) params.key - 1 else null

                        if (response.body() != null) {

                            callback.onResult(response.body()!!.mResults!!, key)

                        }
                    }

                    override fun onFailure(call: Call<MovieApiResponse>, t: Throwable) {
                        print(t)
                    }
                })

    }

    override fun loadAfter(params: PageKeyedDataSource.LoadParams<Long>, callback: PageKeyedDataSource.LoadCallback<Long, Result>) {


        RetrofitClient.instance
                .api
                .getMovies(API_KEY, params.key, LANGUAGE)
                .enqueue(object : Callback<MovieApiResponse> {
                    override fun onResponse(call: Call<MovieApiResponse>, response: Response<MovieApiResponse>) {

                        val key = if (response.body()!!.mTotalPages!! > response.body()!!.mPage!!) params.key + 1 else null

                        if (response.body() != null) {

                            callback.onResult(response.body()!!.mResults!!, key)

                        }
                    }

                    override fun onFailure(call: Call<MovieApiResponse>, t: Throwable) {
                        print(t)
                    }
                })


    }

    companion object {


        private const val FIRST_PAGE: Long = 1
    }
}
